### ISSUE TITLE
About mask encoder

### ISSUE THREAD

**Issue by RichriD** [StavrosOrf/PoweFlowNet#31]
Hello, as the title suggests, I have some questions about the mask encoder.

In the paper, $\boldsymbol{x}_i^0=\boldsymbol{x}_i+\hat{\boldsymbol{m}}_i, \forall i \in \mathcal{N}$ adds the mask to the original data, and the code (MaskEmdMultiMPN) does this as well. However, in Fig. 3 of the paper, it looks like the mask is treated as a new feature of the node, expanding the original node's feature dimension.

I would like to know if these two approaches would affect the results? To me, the method shown in the figure seems more intuitive.

Thank you!

**Comment by StavrosOrf** [comment]
Hello, 

Indeed, we intended for a learnable mask embedding to be added to the initial node features. 

This idea is similar to the way positional embedding works in the transformer Neural Network architectures. By having a mask embedding, we shift the distribution of the input nodes depending on the type of node (load, generator, reference). By shifting the data distributions we expect the following GNN to better learn the desired behavior.

In practice, we tried various architectures and ways to encode the type of node, for example, just by adding an extra node feature as an identifier; however, in the end, the PowerFlowNet (MaskEmdMultiMPN) gave us the best results in terms of training loss.

**Comment by RichriD** [comment]
> Hello,
> 
> Indeed, we intended for a learnable mask embedding to be added to the initial node features.
> 
> This idea is similar to the way positional embedding works in the transformer Neural Network architectures. By having a mask embedding, we shift the distribution of the input nodes depending on the type of node (load, generator, reference). By shifting the data distributions we expect the following GNN to better learn the desired behavior.
> 
> In practice, we tried various architectures and ways to encode the type of node, for example, just by adding an extra node feature as an identifier; however, in the end, the PowerFlowNet (MaskEmdMultiMPN) gave us the best results in terms of training loss.

Thank you very much for your response!

I would like to confirm if my understanding is correct: the current method of embedding the mask yields better performance within the current training framework, which is why this approach has been adopted in the research. At the same time, expanding the mask as a new node feature into the original feature vector is also feasible.

I look forward to your feedback and corrections!

**Comment by StavrosOrf** [comment]
Yes indeed you got it right!

Initially we tried having as node features a 4 number one-hot encoding of the type of node, however we found out it was not as efficient.

It would be interesting to see how having a learnable mask embedding work as additional node features.

**Comment by RichriD** [comment]
Thank you very much for your explanation, it has been incredibly helpful in my understanding of the techniques in your paper. In fact, I am currently looking to apply a similar method to solve the carbon emission flow in power systems and incorporate it into a reinforcement learning framework to address some optimization-related problems.

Thank you again for your clarification

**Comment by StavrosOrf** [comment]
That's good to hear, I wish you best of luck with your research!
